Add info on CVMFS and cvmfsexec to the backfill container doc (#820)
opensciencegrid/docs: docs/resource-sharing/os-backfill-containers.md

Running the Container with Docker
---------------------------------

The Docker image is kept in [DockerHub](https://hub.docker.com/r/opensciencegrid/osgvo-docker-pilot).
In order to successfully start payload jobs:

1. **Configure authentication:**
   Open Science Pool (OSP) administrators can provide the token,
   which you can then pass to the container by volume mounting it as a file under `/etc/condor/tokens-orig.d/`.
   If you are using Docker to launch the container, this is done with the command line flag
   `-v /path/to/token:/etc/condor/tokens-orig.d/flock.opensciencegrid.org`.

>    (Replace `/path/to/token` with the full path to the token you obtained from the OSP administrators.)

Reviewer: You revised the phrasing here but did not drop the parens like in the other location

2. Set `GLIDEIN_Site` and `GLIDEIN_ResourceName` to match the site name and resource name that you registered in Topology,
   respectively.
3. Set the `OSG_SQUID_LOCATION` environment variable to the HTTP address of your preferred Squid instance.
4. _Recommended:_ Enable [CVMFS](#cvmfs) via one of the mechanisms described below.
5. _Optional:_  If you want job I/O to be done in a separate directory outside of the container,
   volume mount the desired directory on the host to `/pilot` inside the container.
   If you are using Docker to launch the container, this is done with the command line flag
   `-v /somelocaldir:/pilot`.
6. _Optional:_ add an expression with the `GLIDEIN_Start_Extra` environment variable to append to the HTCondor `START`
   expression; this limits the pilot to only run certain jobs.

Here is an example invocation using `docker run` by hand:

```

> docker run -it --rm --user osg  \

Reviewer 2: ```suggestion
docker run -it --rm  \
```

`--user osg` breaks Singularity invocations within the container

       --privileged             \
       -v /path/to/token:/etc/condor/tokens-orig.d/flock.opensciencegrid.org \
       -e GLIDEIN_Site="..."                    \
       -e GLIDEIN_ResourceName="..."            \
       -e GLIDEIN_Start_Extra="True"            \
       -e OSG_SQUID_LOCATION="..."              \
       -e CVMFSEXEC_REPOS="                     \
            oasis.opensciencegrid.org           \
            singularity.opensciencegrid.org"    \
       opensciencegrid/osgvo-docker-pilot:release
```

> Replace `/path/to/token` with the location you saved the token obtained from the OSG VO administrators.

Reviewer: You dropped the parens here but did not update the phrasing like in the other location (OSP, etc.)

note that options in environment variables take precedence over options in `/cvmfsexec/default.local`.

You can store the cache outside of the container by volume mounting a directory to `/cvmfs-cache`.

You can store the logs outside of the container by volume mounting a directory to `/cvmfs-logs`.

### Adding CVMFS via bind-mount

As an alternative to using cvmfsexec, you may install CVMFS on the host, and volume mount it into the container.
This will let you avoid running the container in privileged mode.
However, supporting Singularity jobs inside the container will require extra privileges,
namely the capabilities `DAC_OVERRIDE`, `DAC_READ_SEARCH`, `SETGID`, `SETUID`, `SYS_ADMIN`, `SYS_CHROOT`, and `SYS_PTRACE`.

Follow the [installing CVMFS document](../worker-node/install-cvmfs.md) to install CVMFS on the host.

Once you have CVMFS installed and mounted on your host, add `-v /cvmfs:/cvmfs:shared` to your `docker run` invocation.

This is the [example at the top of the page](#running-the-container-with-docker),
modified to volume mount CVMFS instead of using cvmfsexec, and using reduced privileges:

```

> docker run -it --rm --user osg      \

Reviewer 2: ```suggestion
docker run -it --rm                 \
```

`--user osg` breaks Singularity invocations within the container

Author: AFAIK running the container as an arbitrary non-root user was a supported use case. Does this mean we will no longer support that?

Author: https://github.com/opensciencegrid/osgvo-docker-pilot/pull/8#issuecomment-801495129

Reviewer 2: Hrm, good memory so I think we have to better understand why it needs to be a non-root user in the container. I'll re-run some tests but it appears that priv Singularity fails in the `cvmfsexec` case when running as the `osg` user

Reviewer 2: Well duh, priv Singularity invocations will fail as a non-root user: https://github.com/brianhlin/osgvo-docker-pilot/runs/3232846941?check_suite_focus=true#step:5:45

So to allow non-root usage of the container then we need https://opensciencegrid.atlassian.net/browse/SOFTWARE-4725

Reviewer 2: Wait no, I'm not sure that makes sense. The backfill container run through Docker in privileged mode as the `osg` user with CVMFS bind mounted in works fine: https://github.com/brianhlin/osgvo-docker-pilot/runs/3232846885?check_suite_focus=true#step:5:457 but the same command with `cvmfsexec` fails with the error about not finding the setuid: https://github.com/brianhlin/osgvo-docker-pilot/runs/3232846941?check_suite_focus=true#step:5:45. OTOH, running through Docker in privileged mode as the root user with cvmfsexec works fine: https://github.com/brianhlin/osgvo-docker-pilot/runs/3225389938?check_suite_focus=true#step:5:26.

Thoughts @DrDaveD @rynge ?

Reviewer 3: This is still using the old OSG VO wrapper, not glideinWMS, right?  That was the reason why we couldn't yet recommend that people run osgvo-docker-pilot unprivileged yet.  GlideinWMS tries the unprivileged singularity from /cvmfs/oasis.opensciencegrid.org first, I think that's the main difference for us here.

cvmfsexec itself when run as an unprivileged user uses an unprivileged user namespace, so that disables the ability to use setuid within it.  When run as root, sure, you can still run singularity inside it but then it doesn't need setuid because it is running unprivileged.  People definitely won't want to run pilot & therefore user jobs as root, however!

Reviewer 2: Ah that makes sense. Thanks!

Running the Container with Docker
---------------------------------

The Docker image is kept in [DockerHub](https://hub.docker.com/r/opensciencegrid/osgvo-docker-pilot).

In order to successfully start payload jobs:

1. Configure authentication.
   OSG VO administrators can provide the token,
   which you can then pass to the container by volume mounting it as a file under `/etc/condor/tokens-orig.d/`.
   If you are using Docker to launch the container, this is done with the command line flag
   `-v /path/to/token:/etc/condor/tokens-orig.d/flock.opensciencegrid.org`.

>    (Replace `/path/to/token` with the location you saved the token obtained from the OSG VO administrators.)

Author: long live the OSP VO apparently?

>    (Replace `/path/to/token` with the location you saved the token obtained from the OSG VO administrators.)

Author: wait was OSP a typo or on purpose?

>    (Replace `/path/to/token` with the location you saved the token obtained from the OSG VO administrators.)

Reviewer 2: Open Science Pool (OSP) ~ OSG VO. In comms with folks, especially ones new to the OSG, we're trying to avoid the overloaded OSG term

2. Set `GLIDEIN_Site` and `GLIDEIN_ResourceName` to match the site name and resource name you registered in Topology,
   respectively.
3. Set the `OSG_SQUID_LOCATION` environment variable to the HTTP address of your preferred Squid instance.
4. _Recommended:_ Enable [CVMFS](#cvmfs) via one of the mechanisms described below.
5. _Optional:_  If you want job I/O to be done in a separate directory outside of the container,
   volume mount the desired directory on the host to `/pilot` inside the container.
   If you are using Docker to launch the container, this is done with the command line flag

>    `-v /somelocaldir:/pilot`.

Author: If you don't do this, then your execute directory will be in some random place under /var/lib/docker, and user jobs will be competing for space with your containers.

>    `-v /somelocaldir:/pilot`.

Reviewer 2: Makes sense and that's a great reason to mount that dir. I'd say we should recommend this and communicate this exact reasoning!

>    `-v /somelocaldir:/pilot`.

Reviewer 2: I think this is the only thing left that I'd like addressed

>    `-v /somelocaldir:/pilot`.

Author: OK, done -- is my wording OK?

>    `-v /somelocaldir:/pilot`.

Reviewer 2: Yeah, LGTM

6. _Optional:_ add an expression with the `GLIDEIN_Start_Extra` environment variable to append to the HTCondor `START`
   expression; this limits the pilot to only run certain jobs.

Here is an example invocation using `docker run` by hand:

```
docker run -it --rm --user osg  \
       --privileged             \
       -v /path/to/token:/etc/condor/tokens-orig.d/flock.opensciencegrid.org \
       -e GLIDEIN_Site="..."                    \
       -e GLIDEIN_ResourceName="..."            \
       -e GLIDEIN_Start_Extra="True"            \
       -e OSG_SQUID_LOCATION="..."              \
       -e CVMFSEXEC_REPOS="                     \
            oasis.opensciencegrid.org           \

>             singularity.opensciencegrid.org"    \

Author: I made this change because in a previous version of this doc, I _did_ miss it the other way -- and so did my reviewers.

>             singularity.opensciencegrid.org"    \

Author: (Sorry, I missed it in the README file, not this doc. Still.)

       opensciencegrid/osgvo-docker-pilot:release
```

> (Replace `/path/to/token` with the location you saved the token obtained from the OSG VO administrators.)

Reviewer 2: ```suggestion
Replace `/path/to/token` with the location you saved the token obtained from the OSG VO administrators.
```

Privileged mode (`--privileged`) requested in the above `docker run` allows the container

> to mount CVMFS using cvmfsexec (see below) and invoke `singularity` for user jobs.

Reviewer 2: ```suggestion
to mount [CVMFS using cvmfsexec](#adding-cvmfs-using-cvmfsexec) and invoke `singularity` for user jobs.
```

Singularity allows the user to use their own container for their job.

CVMFS
-----

> CVMFS is a read-only remote filesystem that many OSG jobs depend on for software and data.

Reviewer 2: ```suggestion
[CernVM-FS](https://cernvm.cern.ch/fs/) (CVMFS) is a read-only remote filesystem that many OSG jobs depend on for software and data.
```

Supporting CVMFS inside your container will greatly increase the types of OSG jobs you can run.

> There are two methods for adding CVMFS: [enabling cvmfsexec](#adding-cvmfs-using-cvmfsexec),

Reviewer 2: ```suggestion
There are two methods for making CVMFS available in your container: [enabling cvmfsexec](#adding-cvmfs-using-cvmfsexec),
```

or [bind-mounting CVMFS from the host](#adding-cvmfs-via-bind-mount).
Bind-mounting CVMFS will require CVMFS to be installed on the host first,

> but the container will need somewhat fewer privileges.

Reviewer 2: ```suggestion
but the container will need fewer privileges.
```

> ### Adding CVMFS using cvmfsexec

Reviewer 2: I think a link to example invocation above would be good to have somewhere in this section

Reviewer 2: Ah right!

[cvmfsexec](https://github.com/CVMFS/cvmfsexec#readme) is a tool that can be used to mount CVMFS inside the container
without requiring CVMFS to be installed on the host.

To enable cvmfsexec, specify a space-separated list of repos in the `CVMFSEXEC_REPOS` environment variable.
We recommend the following repos:
-   `oasis.opensciencegrid.org`
-   `singularity.opensciencegrid.org`

cvmfsexec has the following system requirements:

-   On EL7, you must have kernel version >= 3.10.0-1127 (run `uname -vr` to check), and user namespaces enabled.
    See step 1 in the
    [Singularity Install document](https://opensciencegrid.org/docs/worker-node/install-singularity/#enabling-unprivileged-singularity)
    for details.

-   On EL8, you must have kernel version >= 4.18 (run `uname -vr` to check).

> See the [cvmfsexec README](https://github.com/cvmfs/cvmfsexec#readme) details.

Reviewer 2: What's in there that's not in this doc? Keep in mind that I want to avoid keeping documentation in the README wherever possible

Author: This is a link to the _cvmfsexec_ README, not the osgvo-docker-pilot README. There are all kinds of technical details in that one that would just clutter up this doc.

Reviewer 2: Ah righto

Note that cvmfsexec will not be run if CVMFS repos are already available in `/cvmfs` via bind-mount,
regardless of the value of `CVMFSEXEC_REPOS`.

Using cvmfsexec takes place in the entrypoint, which means it will still happen
even if you specify a different command to run, such as `bash`.
You can bypass the entrypoint by passing `--entrypoint <cmd>` where `<cmd>` is some different command to run,
e.g. `--entrypoint bash`.
Setting the entrypoint this way clears the command.

#### Additional cvmfsexec configuration

There are several environment variables you can set for cvmfsexec:

-   `CVMFSEXEC_REPOS` - this is a space-separated list of CVMFS repos to mount.
    Leave this blank to disable cvmfsexec.
    OSG jobs frequently use the OASIS repo (`oasis.opensciencegrid.org`) and
    the singularity repo (`singularity.opensciencegrid.org`).

-   `CVMFS_HTTP_PROXY` - this sets the proxy to use for CVMFS; if left blank

>     it will find the best one via WLCG Web Proxy Auto Discovery.

Reviewer 2: Not an issue with the documentation but would it be possible to set the CVMFS proxy location in this order?

1. `CVMFS_HTTP_PROXY`
2. `OSG_SQUID_LOCATION`
3. Auto discovery

Author: I don't know the answer to that. Perhaps @DrDaveD does.

Reviewer 3: The container could do it.  To use auto discovery set `CVMFS_HTTP_PROXY="auto;DIRECT"` because cvmfsexec sets `CVMFS_PAC_URLS` to the appropriate servers.

Reviewer 2: https://opensciencegrid.atlassian.net/browse/SOFTWARE-4691

-   `CVMFS_QUOTA_LIMIT` - the quota limit in MB for CVMFS; leave this blank to
    use the system default (4 GB)

You can add other CVMFS options by bind-mounting a config file over `/cvmfsexec/default.local`;

> note that options in environment variables override options in `/cvmfsexec/default.local`.

Reviewer 2: ```suggestion
note that options in environment variables take precedence over options in `/cvmfsexec/default.local`.
```
